### PR TITLE
SALTO-1966: always have at least an empty type for any deployable instance

### PR DESCRIPTION
**Pull request by noamhammer**
_always have at least an empty type for any deployable instance_

---

_Additional context for reviewer_
None

---
_Release Notes_: 
_Zendesk_support adapter_
* always have at least an empty type for any deployable instance

_Workato adapter_
* always have at least an empty type for any deployable instance

---
_User Notifications_: 
_Zendesk - new empty instances will be added_

### CODE REVIEW

File: packages/adapter-components/src/elements/ducktype/add_remaining_types.ts [hunk 1]
```diff
@@ -0,0 +1,69 @@
...
+*/
+import _ from 'lodash'
+import { ElemID, isObjectType, ObjectType, Element } from '@salto-io/adapter-api'
+import { naclCase, pathNaclCase } from '@salto-io/adapter-utils'
+import { ID_SEPARATOR, toNestedTypeName } from './type_elements'
+import { DATA_FIELD_ENTIRE_OBJECT } from '../../config/transformation'
+import { TypeDuckTypeConfig } from '../../config'
+import { SUBTYPES_PATH, TYPES_PATH } from '../constants'
+
+export const addRemainingTypes = ({
```
Comment (netama) [inline packages/adapter-components/src/elements/ducktype/add_remaining_types.ts:24]: let's have documentation for this function (including the motivation)?

File: packages/adapter-components/src/elements/ducktype/add_remaining_types.ts [hunk 2]
```diff
@@ -0,0 +1,69 @@
...
+    .pickBy(typeConfig => typeConfig.transformation?.sourceTypeName !== undefined)
+    .entries()
+    .map(([typeName, typeConfig]) => [typeConfig?.transformation?.sourceTypeName, typeName])
+    .fromPairs()
+    .value()
+  const typeNames = _(typesConfig)
+    .entries()
+    .flatMap(([typeName, typeConfig]) => {
+      const dataField = typeConfig.transformation?.dataField
+      const standAloneFields = typeConfig.transformation?.standaloneFields
```
Comment (netama) [inline packages/adapter-components/src/elements/ducktype/add_remaining_types.ts:42]: nit:
1. I don't think we need capital a here (standalone is one word)
2. maybe 
```suggestion
      const { dataField, standaloneFields } = typeConfig.transformation ?? {}
```

File: packages/adapter-components/src/elements/ducktype/add_remaining_types.ts [hunk 3]
```diff
@@ -0,0 +1,69 @@
...
+      const nestedFields = [
+        ...((dataField && dataField !== DATA_FIELD_ENTIRE_OBJECT) ? [dataField] : []),
```
Comment (netama) [inline packages/adapter-components/src/elements/ducktype/add_remaining_types.ts:44]: I guess we might have a problem when the datafield is not defined. not sure we have something great to do about it though except for requiring a data field (which I'm not sure we should)

File: packages/adapter-components/src/elements/ducktype/add_remaining_types.ts [hunk 4]
```diff
@@ -0,0 +1,69 @@
...
+    .map(typeName => sourceTypeNameToTypeName[typeName] ?? typeName)
+    .uniq()
+    .value()
+  const existingTypeNames = new Set(elements
+    .filter(isObjectType)
+    .map(e => e.elemID.typeName))
+  const typesToAdd = typeNames
+    .filter(typeName => !existingTypeNames.has(typeName))
+    .map(typeName => {
+      const naclName = naclCase(typeName)
```
Comment (netama) [inline packages/adapter-components/src/elements/ducktype/add_remaining_types.ts:58]: don't we need the `naclCase` earlier (before the lookup)? all existing types are nacl-cased already, no?

File: packages/adapter-components/src/elements/ducktype/add_remaining_types.ts [hunk 5]
```diff
@@ -0,0 +1,69 @@
...
+      return new ObjectType({
```
Comment (netama) [inline packages/adapter-components/src/elements/ducktype/add_remaining_types.ts:59]: can we use a shared function instead of creating a type manually? I assume for example that the logic from https://github.com/salto-io/salto/pull/2713 won't be applied here automatically?

File: packages/adapter-components/src/elements/ducktype/type_elements.ts [hunk 6]
```diff
@@ -23,7 +23,7 @@ import { DuckTypeTransformationConfig, DuckTypeTransformationDefaultConfig, getC
 import { TYPES_PATH, SUBTYPES_PATH } from '../constants'
 import { hideFields } from '../type_elements'
 
-const ID_SEPARATOR = '__'
+export const ID_SEPARATOR = '__'
```
Comment (netama) [inline packages/adapter-components/src/elements/ducktype/type_elements.ts:26]: continuing from the previous comment - hopefully we don't have to export the const if we export a function creating the type

File: packages/zendesk-support-adapter/src/filters/reorder/creator.ts [hunk 7]
```diff
@@ -73,6 +73,7 @@ export const createReorderFilterCreator = (
       { [orderFieldName]: instancesReferences },
       [ZENDESK_SUPPORT, RECORDS_PATH, typeName, typeNameNaclCase],
     )
+    _.remove(elements, element => element.elemID.isEqual(type.elemID))
```
Comment (netama) [inline packages/zendesk-support-adapter/src/filters/reorder/creator.ts:76]: can we do this more explicitly?

**Review by netama — COMMENTED** [review]
added a few comments. can you please share (on the ticket) a before/after repo for workato and zendesk, to make sure the type changes cover everything that we consider risky?

File: packages/adapter-components/src/elements/ducktype/add_remaining_types.ts (hunk 3 above)
Comment (noamhammer) [reply packages/adapter-components/src/elements/ducktype/add_remaining_types.ts:44]: @netama You are correct. what do you suggest me to do here?

File: packages/adapter-components/src/config/ducktype.ts [hunk 8]
```diff
@@ -104,3 +104,9 @@ export const validateFetchConfig = (
     throw Error(`Invalid type names in ${fetchConfigPath}: ${invalidIncludeTypes}`)
   }
 }
+
+export const getTransformationConfigByType = (typesConfig: Record<string, TypeDuckTypeConfig>):
```
Comment (netama) [inline packages/adapter-components/src/config/ducktype.ts:108]: thanks for adding this - 
1. this doesn't look specific for ducktype, can you please move it to a shared location (probably next to `getTypeTransformationConfig`?
2. can you please use this in the relevant places? (I found 3 by searching for the `mapValues` part, two in the ducktype transformer in one for swagger)

File: packages/adapter-components/src/elements/ducktype/add_remaining_types.ts [hunk 9]
```diff
@@ -0,0 +1,73 @@
...
+*/
+import _ from 'lodash'
+import { isObjectType, Element } from '@salto-io/adapter-api'
+import { naclCase } from '@salto-io/adapter-utils'
+import { generateType, toNestedTypeName } from './type_elements'
+import { DATA_FIELD_ENTIRE_OBJECT } from '../../config/transformation'
+import { getTransformationConfigByType, TypeDuckTypeConfig, TypeDuckTypeDefaultsConfig } from '../../config'
+
+/**
+ * Adds the types that supposed to exist but weren't created since they had no instances
```
Comment (netama) [inline packages/adapter-components/src/elements/ducktype/add_remaining_types.ts:24]: nit: `supposed` -> `are supposed`
also: maybe something like 
> Add empty object types for all types that can have instances in the workspace, even if no instances will be created for them in the current fetch.
This is needed because if instances are added / cloned from another environment, they need to have a type in order to be deployed.

File: packages/adapter-components/src/elements/ducktype/add_remaining_types.ts (hunk 3 above)
Comment (netama) [reply packages/adapter-components/src/elements/ducktype/add_remaining_types.ts:49]: discussed IRL - I think maybe we should make  `dataField` required for ducktype adapters that support deploy (but keep it not required for read-only adapters). wdyt?

I think it's not a blocker for merging, but we should add it as part of the MVP because otherwise we might run into unexpected breakages. so let's do this either as part of this ticket or add a separate one so we don't miss it?

File: packages/workato-adapter/test/adapter.test.ts [hunk 10]
```diff
@@ -273,11 +273,8 @@ describe('adapter', () => {
           ),
           elementsSource: buildElementsSourceFromElements([]),
         }).fetch({ progressReporter: { reportProgress: () => null } })
-        expect(elements).toHaveLength(7)
-        expect(elements.filter(isObjectType)).toHaveLength(1)
-        expect(elements.filter(isInstanceElement)).toHaveLength(6)
-        expect(elements.map(e => e.elemID.getFullName()).sort()).toEqual([
-          'workato.connection',
+        const intances = elements.filter(isInstanceElement)
```
Comment (netama) [inline packages/workato-adapter/test/adapter.test.ts:276]: why was this changed? can you please add back the object types?

File: packages/zendesk-support-adapter/src/filters/reorder/creator.ts [hunk 11]
```diff
@@ -74,6 +74,9 @@ export const createReorderFilterCreator = (
       { [orderFieldName]: instancesReferences },
       [ZENDESK_SUPPORT, RECORDS_PATH, typeName, typeNameNaclCase],
     )
+    // Those types are already exist since we added the empty version of them
```
Comment (netama) [inline packages/zendesk-support-adapter/src/filters/reorder/creator.ts:77]: nit: remove `are`

**Review by netama — APPROVED** [review]
LGTM with a few remaining comments (and one change that we can do separately)

File: packages/workato-adapter/test/adapter.test.ts (hunk 10 above)
Comment (noamhammer) [reply packages/workato-adapter/test/adapter.test.ts:276]: @netama I can add them back but please note that now we also create at least one obj type for each type, so we will have all the other types as well

File: packages/adapter-components/src/elements/ducktype/add_remaining_types.ts (hunk 3 above)
Comment (noamhammer) [reply packages/adapter-components/src/elements/ducktype/add_remaining_types.ts:52]: https://salto-io.atlassian.net/browse/SALTO-2007
